components/capi/src/custom_writeable.rs

```diff
@@ -0,0 +1,97 @@
+use std::ffi::c_void;
+use std::{fmt, ptr};
+
+#[repr(C)]
+/// An object that can one can write UTF-8 strings to
+///
+/// This allows the C API to write to arbitrary kinds of objects, for example a
+/// C++ std::string or a char buffer.
+///
+/// [`icu4x_simple_writeable()`] can be used to write to a fixed-size char buffer.
+///
+/// May be extended in the future to support further invariants
+///
+/// # Safety invariants:
+///  - `flush()` and `grow()` will be passed `data` and the value should be valid for that
+///     `data` may be  null, however `flush()` and `grow()` must then be ready to receive it
+///  - `buf` must be `cap` bytes long
+///  - `grow()` must either return null or a valid buffer of at least the requested buffer size
+///  - Rust code must call `ICU4XCustomWriteable::flush()` before releasing to C
+pub struct ICU4XCustomWriteable {
+    /// Pointer to the actual object. While we're writing, we will write
+    /// directly to `buf` without updating `data`'s state, this pointer exists so that
+    /// `grow()` and `flush()` can get access to the full object on the foreign side
+    data: *mut c_void,
+    /// The buffer to write directly to
+    buf: *mut u8,
+    /// The current filled size of the buffer
+    len: usize,
+    /// The current capacity of the buffer
+    cap: usize,
```

Comment:
Is it correct to say that these are "Rust-only" fields?  Should we make them "hidden" from C?

Reply:
No, they all get initialized on the C side.